replica/table.cc

```diff
@@ -106,11 +106,6 @@ table::make_sstable_reader(schema_ptr s,
     // consequence, fast_forward_to() will *NOT* work on the result,
     // regardless of what the fwd_mr parameter says.
     if (pr.is_singular() && pr.start()->value().has_key()) {
-        const dht::ring_position& pos = pr.start()->value();
-        if (_erm->shard_of(*s, pos.token()) != this_shard_id()) {
-            return make_empty_flat_reader_v2(s, std::move(permit)); // range doesn't belong to this shard
-        }
-
```

Comment:
The optimization was introduced in 66dfeb33d7812b7e49cd01e4b2ebd60cc80b51b0, without noting when it applies. Agree it can be removed, storage_proxy should route to the correct shard even with vnodes.

Reply:
not relevant to single key reads but while we're here, I suggest we patch cql layer (in partition key restriction or something) to stop using (-inf, +inf) in full scans. as there will be an extra attempt to read on range (max_token, +inf) which is not owned by any tablet (useless round trip and also inviting for problems if replica incorrectly wraps around from max to min). is there any historical reason we do it? why not [first_token = next(min_token), max_token] instead when there's no pkey restriction on the scan?

Reply:
Why is there a round trip after max_token? It will be squelched in the coordinator.

Reply:
> The optimization was introduced in [66dfeb3](https://github.com/scylladb/scylladb/commit/66dfeb33d7812b7e49cd01e4b2ebd60cc80b51b0), without noting when it applies. Agree it can be removed, storage_proxy should route to the correct shard even with vnodes.

It's explained in the merge commit (5d9222d935c1f5dfcc18649a35246987e6541a58):

> 
> Merge "Filter sstable data not belonging to current shard" from Tomasz
> 
> "We don't want multiple shards to respond with the same data. Higher level code
> assumes that shard data is non-overlapping. It's cheaper to drop duplicates as
> soon as possible. Memtable reader for example will never have overlapping
> data, so cache hitting queries will never need to pay for this. Compaction
> process may also rely on this."

It was needed when sstables could be shared among shards.